src/main/java/org/opensearch/flowframework/workflow/DeleteSearchPipelineStep.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+package org.opensearch.flowframework.workflow;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.ExceptionsHelper;
+import org.opensearch.action.search.DeleteSearchPipelineRequest;
+import org.opensearch.action.support.PlainActionFuture;
+import org.opensearch.client.Client;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.flowframework.exception.WorkflowStepException;
+import org.opensearch.flowframework.util.ParseUtils;
+
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+import static org.opensearch.flowframework.common.WorkflowResources.PIPELINE_ID;
+import static org.opensearch.flowframework.exception.WorkflowStepException.getSafeException;
+
+/**
+ * Step to delete a search pipeline
+ */
+public class DeleteSearchPipelineStep implements WorkflowStep {
```

Comment:
Looks like except request the code is same for both the steps. We can create an abstract class for both DeleteSearch and DeleteIngest Pipeline step.

Reply:
Yeah, we could, but I don't think there's any benefit.

I did look at the abstract class for create when writing this and considered it.

Unfortunately the Create abstract class breaks the beauty of inheritance by doing this silly stuff internally, which it should not have done. (If you insist on combining the rest of the method, just call an abstract method here and implement it on the subclass.)
```json
            if (pipelineToBeCreated.equals(CreateSearchPipelineStep.NAME)) {
                PutSearchPipelineRequest putSearchPipelineRequest = new PutSearchPipelineRequest(
                    pipelineId,
                    configurationsBytes,
                    XContentType.JSON
                );
                clusterAdminClient.putSearchPipeline(putSearchPipelineRequest, putPipelineActionListener);
            } else {
                PutPipelineRequest putPipelineRequest = new PutPipelineRequest(pipelineId, configurationsBytes, XContentType.JSON);
                clusterAdminClient.putPipeline(putPipelineRequest, putPipelineActionListener);
            }
```

This appears to make some sense for the create step because of the overlap in code, but I'd propose that even that is a misguided refactoring.  All the "failed to update resource in state index" should be handled inside the `FlowFrameworkIndicesHandler` class.  That bit of code in the abstract class is [duplicated in almost every resource-creating step](https://github.com/search?q=repo%3Aopensearch-project%2Fflow-framework%20flowFrameworkIndicesHandler.updateResourceInStateIndex(&type=code).   So there's some refactoring to be done, but not into an abstract parent class!

Subtracting out the `FlowFrameworkIndicesHandler` code and the static constants that belong in each subclass, there's really not much left to put in the parent.  

And the delete steps are deprovision steps, so they don't even use that code; the deprovision API handles updating the state index.

The only real duplicated code I see is fetching the previous inputs via the ParseUtil (one method) and then pulling the PIPELINE_ID out of the map.  Perhaps the `client` instance field that could be put in the superclass.   That's essentially two lines of duplicated code (or 3 if you move the client field).   The rest is unique (static constants in the class, a different request class, and a different cluster admin client call. (Sure, we could maybe extract out the AcknowledgedResponse handling but that's still just a few lines of code.)

I don't think an abstract class is useful at all in this case, and may not be very useful in the Create case either once the indices handler class handles all that repeat code and the incorrect subclass implementations are properly removed.

Reply:
Agreed. Since our lot of code is getting duplicated in multiple workflow steps, we can refactor or pull that code out in the parent abstract class. Can we have an issue for refactoring?